shub/utils.py

```diff
@@ -143,17 +143,19 @@ def _check_deploy_files_size(files):
 
 def write_and_echo_logs(keep_log, last_logs, rsp, verbose):
     """It will write logs to temporal file and echo if verbose is True."""
-    with NamedTemporaryFile(prefix='shub_deploy_', suffix='.log',
-                            delete=(not keep_log)) as log_file:
-        for line in rsp.iter_lines():
-            if verbose:
-                click.echo(line)
-            last_logs.append(line)
-            log_file.write(line + b'\n')
+    log_contents = b""
+    for line in rsp.iter_lines():
+        if verbose:
+            click.echo(line)
+        last_logs.append(line)
+        log_contents += line + b'\n'
+    deployed = _is_deploy_successful(last_logs)
+    keep_log = echo_short_log_if_deployed(deployed, last_logs, keep_log, verbose)
 
-        deployed = _is_deploy_successful(last_logs)
-        echo_short_log_if_deployed(deployed, last_logs, log_file, verbose)
-        if not log_file.delete:
+    with NamedTemporaryFile(prefix='shub_deploy_', suffix='.log',
+                            delete=not keep_log) as log_file:
+        log_file.write(log_contents)
+        if keep_log:
```

Comment:
How does this change relate to 3.12?

Reply:
> `NamedTemporaryFile` no longer has a `delete` attribute

Here `write_and_echo_logs` tries to get that attribute:

https://github.com/scrapinghub/shub/blob/0f3087f32b81c4ac6a966bb23f0c2414c349eff6/shub/utils.py#L156

On 3.12, that throws `AttributeError: '_io.BufferedRandom' object has no attribute 'delete'`. Also `echo_short_log_if_deployed` sets `delete` when the deploy failed which no longer makes sense (in fact, this didn't have the intended effect of keeping the file since `__exit__` didn't respect it, only [`_closer.delete`](https://github.com/python/cpython/blob/3.11/Lib/tempfile.py#L509)):

https://github.com/scrapinghub/shub/blob/0f3087f32b81c4ac6a966bb23f0c2414c349eff6/shub/utils.py#L171

So instead, this change just waits to construct the temp file until the final value for `delete` is determined